src/Features/Core/Portable/EmbeddedLanguages/StackFrame/StackFrameLexer.cs

```diff
@@ -0,0 +1,368 @@
+        /// </summary>
+        public StackFrameTrivia? ScanTrailingTrivia()
+        {
+            if (Position == Text.Length)
+            {
+                return null;
+            }
+
+            var length = Text.Length - Position;
+            return CreateTrivia(StackFrameKind.TrailingTrivia, Text.GetSubSequence(new TextSpan(Position - 1, length)));
+        }
+
+        public StackFrameToken? ScanIdentifier()
+        {
+            if (Position == Text.Length)
+            {
+                return null;
+            }
+
+            var startPosition = Position;
+
+            var ch = CurrentChar;
+            Position++;
+
+            if (!UnicodeCharacterUtilities.IsIdentifierStartCharacter((char)ch.Value))
+            {
+                return null;
+            }
+
+            while (UnicodeCharacterUtilities.IsIdentifierPartCharacter((char)ch.Value))
+            {
+                ch = CurrentChar;
+                Position++;
+            }
+
+            var identifierEnd = Position - 1;
+            var identifierSpan = new TextSpan(startPosition, identifierEnd - startPosition);
+            var identifier = CreateToken(StackFrameKind.IdentifierToken, Text.GetSubSequence(identifierSpan));
+            return identifier;
+        }
+
+        public StackFrameToken ScanTypeArity()
+        {
+            if (Position == Text.Length)
+            {
+                return default;
+            }
+
+            var startPosition = Position;
+            var ch = CurrentChar;
+            Position++;
+
+            while (IsNumber(ch))
+            {
+                ch = CurrentChar;
+                Position++;
+            }
+
+            var aritySpan = new TextSpan(startPosition, (Position - 1) - startPosition);
+            var arityToken = CreateToken(StackFrameKind.TextToken, Text.GetSubSequence(aritySpan));
+            return arityToken;
+        }
+
+        internal ImmutableArray<StackFrameToken> ScanArrayBrackets()
+        {
+            if (Position == Text.Length)
+            {
+                return default;
+            }
+
+            using var _ = ArrayBuilder<StackFrameToken>.GetInstance(out var builder);
+
+            while (Position < Text.Length)
+            {
+                // Use PreviousCharAsToken here because we expect the 
+                // start of the brackets to be the previous token that started
+                // the need to parse for an array
+                var token = PreviousCharAsToken();
+                if (!IsArrayBracket(token))
+                {
+                    break;
+                }
+
+                Position++;
+                builder.Add(token);
+            }
+
+            Debug.Assert(builder.Count >= 2);
+
+            return builder.ToImmutable();
+
+            static bool IsArrayBracket(StackFrameToken token)
+                => token.Kind is StackFrameKind.OpenBracketToken or StackFrameKind.CloseBracketToken;
+        }
```

Comment:
i don't really get this.  something feels fishy about how the char/token handling is happening.